components/movementsensor/adxl345/adxl345.go

```diff
@@ -232,6 +331,123 @@ func (adxl *adxl345) writeByte(ctx context.Context, register, value byte) error
 	return handle.WriteByteData(ctx, register, value)
 }
 
+func getByteFromInterrupts(interrupts []string) byte {
+	var register byte
+	for _, interrupt := range interrupts {
+		register += interruptBitPosition[interrupt]
+	}
+	return register
+}
+
+func (adxl *adxl345) enableInterrupts(ctx context.Context) error {
+	register := getByteFromInterrupts(adxl.interruptsEnabled)
+	return adxl.writeByte(ctx, IntEnable, register)
+}
+
+func (adxl *adxl345) setRelevantRegisters(ctx context.Context) error {
+	if len(adxl.interruptsEnabled) == 0 {
+		return nil
+	}
+	relevantRegisters := adxl.getRelevantRegisterValues()
+	for key := range defaultRegisterValues {
+		val, ok := relevantRegisters[key]
+		if ok {
+			err := adxl.writeByte(ctx, key, val)
+			if err != nil {
+				return errors.Errorf("Unable to write byte at register %x", key)
+			}
+		} else {
+			err := adxl.writeByte(ctx, key, defaultRegisterValues[key])
+			if err != nil {
+				return errors.Errorf("Unable to write byte at register %x", key)
+			}
+		}
+	}
+	return nil
+}
+
+func (adxl *adxl345) getAxes(x, y, z bool) byte {
+	var tapAxes byte
+	if x {
+		tapAxes += xBit
+	}
+	if y {
+		tapAxes += yBit
+	}
+	if z {
+		tapAxes += zBit
+	}
+	return tapAxes
+}
+
+func (adxl *adxl345) getRelevantRegisterValues() map[byte]byte {
+	relevantRegisterValues := map[byte]byte{}
+
+	tapAxesSpecified := adxl.tapX || adxl.tapY || adxl.tapZ
+	if tapAxesSpecified {
+		relevantRegisterValues[TapAxes] = adxl.getAxes(adxl.tapX, adxl.tapY, adxl.tapZ)
+	}
+
+	var actInactAxes byte
+	activityAxesSpecified := adxl.actX || adxl.actY || adxl.actZ
+	if activityAxesSpecified {
+		actInactAxes = adxl.getAxes(adxl.actX, adxl.actY, adxl.actZ)
+	}
+	inactivityAxesSpecified := adxl.inactX || adxl.inactY || adxl.inactZ
+	if inactivityAxesSpecified {
+		actInactAxes += (adxl.getAxes(adxl.inactX, adxl.inactY, adxl.inactZ) << 4)
+	}
+	if activityAxesSpecified || inactivityAxesSpecified {
+		relevantRegisterValues[ActInactCtl] = actInactAxes
+	}
+
+	if adxl.threshTap != 0 {
+		relevantRegisterValues[ThreshTap] = adxl.threshTap
+	}
+	if adxl.dur != 0 {
+		relevantRegisterValues[Dur] = adxl.dur
+	}
+	if adxl.latent != 0 {
+		relevantRegisterValues[Latent] = adxl.latent
+	}
+	if adxl.window != 0 {
+		relevantRegisterValues[Window] = adxl.window
+	}
+	if adxl.threshFf != 0 {
+		relevantRegisterValues[ThreshFf] = adxl.threshFf
+	}
+	if adxl.timeFf != 0 {
+		relevantRegisterValues[TimeFf] = adxl.timeFf
+	}
+	if adxl.threshAct != 0 {
+		relevantRegisterValues[ThreshAct] = adxl.threshAct
+	}
+	if adxl.threshInact != 0 {
+		relevantRegisterValues[ThreshInact] = adxl.threshInact
+	}
+	if adxl.timeInact != 0 {
+		relevantRegisterValues[TimeInact] = adxl.timeInact
+	}
+	if adxl.actInactCtl != 0 {
+		relevantRegisterValues[ActInactCtl] = adxl.timeFf
+	}
+	return relevantRegisterValues
+}
+
+func (adxl *adxl345) readInterrupts(ctx context.Context) {
+	interuptEnabledRegister := getByteFromInterrupts(adxl.interruptsEnabled)
```

Comment:
The only thing we ever use `adxl.interruptsEnabled` for is calling `getByteFromInterrupts`, and that's a fairly slow  function (looking up strings in a map). How about just calling it once in `NewAdxl345` and storing that byte, instead of recomputing it every time?

Reply:
This was sort of to protect against any changes made to `interruptsEnabled` throughout its use. However I am not sure how one would do that. If we do not expect a user to update the `interruptsEnabled` I think we can just use `adxl.interruptsEnabled` here. Am I understanding what you mean correctly?